### ISSUE TITLE
Question about mOffsetMatrix.

### ISSUE THREAD

**Issue by malortie** [assimp/assimp#2773]
Hello,

For the last months, I have been coding support for Half-Life 1 MDL files for Assimp. The work is almost complete but I still have a few questions (will be posted in separate issues). One of them is regarding how to calculate `mOffsetMatrix`.

According to Assimp's documentation, mOffsetMatrix represents the inverse bind pose matrix. I was not entirely confident with my understanding on this, then I came along this page:

https://stackoverflow.com/questions/17127994/opengl-bone-animation-why-do-i-need-inverse-of-bind-pose-when-working-with-gp

According to the answerer, the inverse bind pose matrix 'undoes' the changes that have been applied to the bind pose. After reading, I believe I understand how to calculate the inverse bind pose matrix. I will explain with an example and would like you to correct me if I am wrong.

#### Example

Imagine we have a set of bones, let's call it (**S**). The set (**S**) are the bones with no transformation applied. In short, (**S**) are all bones in 'local space'.

Now if we take every bone of (**S**) and multiply each of them by their parent matrix, we get the set (**T**), which represents all bones in 'bind pose' in 'model space'. The matrix (**M**) is the matrix that converts
 `S -> T`.

So following what the answerer said, the inverse bind pose matrix (**N**) is supposed to undo changes applied to the bind pose, so it should technically represent the inverse of each bind pose matrix. This means `N  =  M^-1  <=>  T -> S  <=>  (S -> T)^-1.`

It goes out saying that If I multiply each inverse bind pose matrix by it's associated bind pose bone matrix of set (**T**), then I should get the identity matrix. 

Is this correct?

Thanks.

**Comment by bztsrc** [comment]
Maybe this helps: imagine that there's a humanoid model. It consist of 6 meshes: head, body, arms, legs; each associated with exactly one bone. With the bind-pose, these are placed properly, head on the top, arms on the two sides of the body, legs below. This is bind-pose. After applying mOffsetMatrix, each mesh will be transformed to a bone local space, that is, all meshes would have the zerus point of the coordinate system; thus if you visualize this, the parts would be shown overlapping around the origin. If you concatenate the transformation matrices of the bone nodes along their path, you'll get a bone transformation matrix. Multiplying the meshes with these (each with its corresponding bone's matrix) would bring all meshes back into bind-pose.

Unfortunately it is not this simple, because matrix multiplication is not commutative. For the sake of the example, assume that we have 2 bones per arms. So to get the bind-pose, you have to multiply
B1 x B2 x B3 x O x M
where B1 is the mTransformation matrix of the top of the bone hierarchy (let's say body), B2 is it's child (let's say upper arm), and B3 is the leaf in the hierarchy (lower arm). O is the mOffsetMatrix, and M is the mTranformation matrix of the mesh. For bind-pose, the inverse of B1 x B2 x B3 is O, thus the above gives
B1 x B2 x B3 x O x M = I x M = M
where I is the identity matrix, meaning you only apply the M mesh transform.
Now for an animation-pose, let's say in an aiNodeAnim you've position, rotation, scaling for B2 bone for a frame. You calculate the mTransformation matrix for those aiVector3Ds and aiQuaternion, and you replace B2's mTranformation with that. After that,
B1 x B2' x B3 x O x M
will get you the animation-pose of the mesh at that frame. Here the inverse of B1 x B2' x B3 does not equal to O, therefore it moves the mesh.
Hope this makes sense to you.

**Comment by malortie** [comment]
@bztsrc 

Thank you for your reply.

I believe I now understand. The point of mOffsetMatrix is to unapply all parent transformations of the bone in 'bind pose' so that the bone gets back to the same position when the skeleton was 'bound'. It would overlap the origin, as you stated. In the case of a 'bind pose', the resulting matrix is the identity matrix since the two matrices mOffsetMatrix * (absolute parent transform) cancel each other, but for an animated bone, this could be different depending on the pose of the animated bone.

A good test to make would be the following:

1. Calculate the bind pose of each bone.
2. Get the mOffsetMatrix of each bone (from the model loader).
3. Transform (in the correct order) each bone (in bind pose) 's parent **absolute** transformation by mOffsetMatrix.
4. Test that the result is the identity matrix. Since we made no changes, then those two matrices cancel out.

Is this correct?

Thanks.

**Comment by bztsrc** [comment]
Almost correct.

First, "mOffsetMatrix * (absolute parent transform)" is not the same as "(absolute parent transform) * mOffsetMatrix". You get used to the fact the multiplication is commutative, eg. you can replace the parts. However this is not true for matrix multiplication, there the order matters.

Second, mOffsetMatrix is the inverse of the accumulated bind-pose bone (the absolute parent transform). The vertices are already in this pose (let's assume for a moment mesh mTransformation is the identity matrix, and see my Note1), so if you transform them by the bone mTranformation (absolute parent transform) again, they would scatter away. The mOffsetMatrix most likely counteracts this (I say most likely, see Note1). One could wonder what's the point, and the answer is, when you animate, you'll get a new bone mTransformation (probably for only one or two bones, not necessarily all the bones), so you have to recalculate the accumulated bone matrix for that bone, then the resulting absolute mTransformation will differ from the bind-pose's accumulated absolute matrix. That difference is what makes the mesh "move" on a particular frame.

Note1: another reason, mOffsetMatrix not necessarily counteracts the mesh's pose for the bind-pose. For example, a mesh of a knight could have his sword next to the knight, and the bone's mOffsetMatrix could transform the sword into the knight's hand. In this case the mesh-pose and the bind-pose would be different (which is not uncommon, because it's easier for designers to edit multiple parts next to each other, and only move everything into place in the bind-pose).

The purpose of the mesh mTransform (we haven't spoken about that yet), is to make the models more compact. For example, with a humanoid model, one could store two separate meshes for the arms. Or, one could only store the right arm, and use a mirroring mesh transform to get the left arm's vertices. A castle for example could store only one tower, and use 4 mesh transformations to place the same  tower on the four edges of the castle. Of course with those mirrored or repositioned meshes their mOffsetMatrix will differ. You could use the same bones though, however it is more likely you want separate bones for them, because for example you'll want to animate the two arms differently, and for that you'll need to separate absolute parent bone transform as well.

About your test, I think points 1. and 3. are the same, otherwise yes. The accumulated bone transform (which is multiplying all bone mTransformations along the bone's path) should be equal to the inverse of the mOffsetMatrix.

Cheers,
bzt

**Comment by malortie** [comment]
@bztsrc 

Thank you for your reply, as well as for your patience.

> You get used to the fact the multiplication is commutative, eg. you can replace the parts. However this is not true for matrix multiplication, there the order matters.

In regards to matrix multiplication being not commutative, I was already aware that it is not commutative and the order of the multiplication matters.

I think your example with the knight and the sword helped me understand how mOffsetMatrix is represented. See below.

1. Knight standing in T pose with his sword in his hand. The bind pose is recorded. At this point, the inverse bind pose matrix is the inverse of each bone's **parent absolute transform** and the sword's bone offset matrix is the identity matrix.

2. Designer moves the bone of the sword away from the hand of the knight to make it easier to edit. The sword mesh follows its bone of course. From now, the sword's bone offset matrix is no longer an identity matrix. It is the transformation matrix to combine with the bind pose, for which the combined matrix once inverted brings the sword bone back in the hand of the knight at bind pose. More technically, it is **(OffsetMatrix x BindPoseParentAbsoluteTransformMatrix).Inverse()**. In Assimp, the **mOffsetMatrix** represents **(OffsetMatrix x BindPoseParentAbsoluteTransformMatrix).Inverse()**.

3. After he finished editing the sword, the designer decides to move the bone of the sword back in the same position as it was in bind pose (See 1.). The sword bone's offset is again an identity matrix, but the inverse bind pose matrix always remains the inverse of the parent's absolute transform. He saves the file, and in this case we lose no information.

    OR

    The designer decides not to move the bone back into its original bind pose position. Based on the exporter/model compiler, the offset matrix may or not be written. In the case we could lose the offset information.

I think I also understand now why Assimp stores an mOffsetMatrix per mesh. When the mOffsetMatrix is known in advance, this is useful and allows to correctly bring the bone back to bind pose, or in bone local space when mOffsetMatrix is applied directly to the bind pose. But often times, the offset 'matrix' information is lost, so there is no way to know if the bone was offset after the file was saved. Because the Assimp importers do not always have this info, they assume that the bones are positioned in bind pose and that there is no offset, and only apply the 'inverse bind pose' matrix.

Is this correct?

If this is correct, please do not close this issue because I might have one or more questions.

Thanks.

**Comment by bztsrc** [comment]
Have no worries, I used to be a teacher :-) Plus this information might be useful to others too.

About 1., correct, but it is only the identity matrix if the mesh is already stored in the bind pose. (If the mesh is not in bind-pose, then the parent absolute transform matrix is not identity matrix, therefore it's inverse, the mOffsetMatrix isn't either.)

About 2., also correct, but meshes are not necessarily in bind-pose. When a designer opens up a model in a modeller, they see the mesh pose (that's not exactly a pose, but I don't have better word for it). Let's say mesh space. In mesh space the sword is next to the knight, and the sword and the knight is stored as two different meshes. Then the designer applies the armature / skeleton (whatever their software calls bone transformations) to display the bind-pose, and the sword moves into the knight's hand. 

About 3., when they save the model, both information is stored: the mesh is saved by vertices in mesh-space, and the bone structure as an armature / skeleton / joint list etc. (depending on format). Maybe through an example it would be easier to understand, read http://tfc.duke.free.fr/coding/md5-specs-en.html With the MD5 format, meshes and animations are in separate files, that's what makes it a good example. The .md5mesh contains the mesh. It also has a joint list (which is a sort-of bone-list but in world-space), but the only purpose is so that you can calculate the mOffsetMatrices. Here the mesh pose (what's listed in mesh{ vert } blocks) only differs from the bind-pose (mesh { weights } array applied) if more bones are influencing the same vertices. What important is, the mesh-space vertices and the bind-pose are stored independently (so that the designer don't need to take the sword out and put it back into the knight's hands, as both information is in the model file at the same time, it's just a matter whether bone-transform is enabled or not). Now the .m5anim files stores the real bone-hierarchy, which moves the meshes into animation-poses. It is not impossible, but uncommon that the joint list in the .md5mesh file matches any joint list in an .md5anim file bit by bit. Most commonly it's only used to identify which animations can be loaded for that particular mesh.

The mOffsetMatrix is not stored per mesh, it is stored per bone. So for example knight is stored in mesh 1, with bones 1-7, and the sword in mesh 2 with bone 8. Then for the sake of the example, let's say the knight's mesh is already in bind-pose (so absolute parent matrices and mOffsetMatrices for bones 1-7 and mesh 1 are identity matrices), but the mTransformation of bone 8 and mOffsetMatrix for bone 8 in mesh 2 is not (as bone 8 has to be moved and rotated to transform the sword's mesh from mesh-space into the knight's hand in bone-pose). I hope this makes now sense to you.

The offsetmatrix is always known in advance. If there's a skeleton (armature / joint list etc.), then you can calculate the absolute parent matrices, and there, with an inverse, you'll get the offsetmatrix. If there's no bone information in the model, then you can assume both absolute parent matrix and offsetmatrix is the identity matrix, and that the mesh-space is the same as the bind-pose. Either way, you'll have the mOffsetMatrix in advance. There's no need to store the mOffsetMatrix in the file really (the .md5mesh does not store it for example), but many model formats do to speed up model loading (however that could lead to data inconsistency: if the inverse of bind-pose absolute parent transforms differ to the offsetmatrix that you loaded, then you have to decide which one you trust).

With an Assimp example: let's say you iterate through mMeshes, and only apply the mesh mTransformation. You'll get the mesh-space representation of the model (sword is next to the knight). If you also apply mOffsetMatrix and the absolute parent matrix in bone aiNodes, then you'll get the bind-pose (sword in knight's hand).

An actual example https://github.com/ccxvii/asstools/blob/master/assview.c here in line 402 the drawscene() starts with an identity matrix, then it calls drawnode() in line 367, which applies the mTransformation matrix and calls itself recursively (this is to take care of the M part of the Bn x O x M equation). If the node has a mesh, then it is drawn by drawmesh(), which just draws the mesh. See the "if" in line 377, that's what converts from mesh-space into bind-pose. That calculation is either skipped (Bn and O both being identity matrix, so only M needed), or the transformmesh() in line 257 already did that conversion. And that's our last step, transformmesh() is the function that calculates absolute parent matrices (Bn) and multiplies vertices by that and O to move the mesh vertices from mesh-space into bind-pose.

Cheers,
bzt

**Comment by bztsrc** [comment]
Just a plus note to be complete, as I've said, transformmesh() converts from mesh-space into bind-pose, but for the animation, animatescene() in line 487 replaces some of the bone mTransformations in aiNodes first. Then animatescene() calls transformmesh() for all meshes, only this time that will move the vertices from mesh-space to animation-pose (because some of the bone node's mTransformations have been altered).

Cheers,
bzt

**Comment by malortie** [comment]
Thanks.

I launched the example (**assview.c**) you mentioned and inspected the rendering process. Now, I see how the inverse bind pose matrix (**mOffsetMatrix**) affects everything. I tested all your scenarios and came up two others (3. and 4.). I will explain my understanding of offset and inverse bind pose transformations below.  Also, because you explained that transformation matrices from animation nodes replace the original mTransformation of their associated nodes, and because I am confident with my understanding on this, the following examples are only about getting vertices of each mesh into bind pose with **no** animation. This is also to keep things simple just to ensure that we understand each other.

First, let's review the math formula you mentioned in your first answer:

**B1 x B2 x B3 x O x M**

For the purpose of explaining, I will temporarily remove (**O**) from the expression above. Therefore we get:

**B1 x B2 x B3 x M**

For readability, let **P = (B1 x B2 x B3)**. This stands for "Parent" absolute transformation. Therefore we get:

**P** x **M**

The goal is to have (**M**) in bind pose at the end of the process.

1. In the case where (**M**), is positioned in the same way as the bind pose and knowing that we have (**P**), that is an incoming transformation to be applied to (**M**), we do not want to apply (**P**) because it would be applying the absolute parent transformation twice. To prevent this, we need to cancel the incoming transformation (**P**). We do this by introducing a new transformation matrix (**O**), which will cancel the incoming transformation. This gives us:

    **P x O x M   <=>  P x (P^-1) x M  =>  I X M  =>  M**

    Therefore, we are left with (**M**). No vertices have moved and (**M**) remains unaffected and therefore, still in bind pose, as we wanted.

2. In the case where (**M**), in mesh space, is not positioned the same way as the bind pose, we want to have a way to move (**M**) in the desired position in bind pose. Let's examine the expression:

    **P x O x M**

    (**O**) is the exact same matrix as (**O**) in (1.). With (**O**), we can cancel the incoming transformation (**P**), which we do not want to apply twice. However, we want to move (**M**) to the desired position in bind pose. We need an additional matrix to do that.

    **P x O x ? x M**

    Assuming we know the matrix that represents the extra offset transformation of (**M**), in mesh space, with respect to the desired position then we need a matrix to apply the 'opposite' effect. This will be the inverse of the extra offset transformation, which we will name (**A**). Therefore, we have:

    **P x O x A x M**

    If we evaluate the expression above, we have:

    **P x O x A x M  =>  I x A x M  => A x M**

    This will position and orient (**M**) correctly as in the bind pose  .e.g: The sword into the knight's hand. 

3. Imagine now a model with all meshes modeled the same way as in bind pose and where all bones, originally positioned in bind pose, were saved relative to their parent. In short, bones and meshes would be overlapping around the origin. Assuming there are no extra offsets in this case, and we want the meshes and bones to be in bind pose, then we need to apply (**P**). Let's look at the formula:

    **P x O x M**

    In this case, (**O**) must not cancel the incoming transformation (**P**) because the bones are in local space and we want to move them into bind pose, so we make (**O**) the identity matrix. Therefore we have:

    **P x O x M  =>  P x I x M  => P x M**

    This will transform all bones, currently relative to their parent, into bind pose as we wanted. Because all meshes are relative to the joints and because there are no offsets, the meshes will then move in bind pose, as we wanted.

4. Imagine the same as (3.) except there is a mesh that contains an extra offset transformation. As explained, bones are in space relative to their parents (local space), and so are the meshes. One needs to be cautious.

    **P x O x A x M**

    In this case, (O) needs to be an identity matrix because it must not cancel (**P**). What is left to deal with is (A). Since we are currently in local space, then (**A**) must bring (**M**) to the correct position, in local space before applying the incoming transformation (**P**). Therefore we have:

    **P x O x A x M  =>  P x I x A x M  => P x A x M**

    where (**A**) is the inverse of the extra offset matrix to be applied in local space. This allows us to move (**M**) to bind pose, as we wanted.

**Note**: In the case where a designer works with a model and offsets some meshes away from the bind pose, it's possible via export tools to calculate the final bind pose matrix and apply to this mesh directly, as well as the bind pose matrix for each bone, before saving the model to a file.  In this case and because, the mesh is stored in bind pose, then we just need to apply (1.).

**Note 2**: The extra offset transformation matrix can either be in bind pose space or local space. It is important to use the correct one or transform it to the appropriate space before it can be applied.

**Note 3**: In the event where there is at least one extra offset transformation in the model, and no information about this is present, and the plugin does not export or store info for this extra offset, then there is no way to guess/find the missing extra offset.

There are many more cases that can arise, but those are often time format dependent and the only thing that can be done is to understand how vertices,joints/bones are stored and compute the appropriate matrix to ensure they are correctly transformed into bind pose after the incoming transformation (**P**) is applied.

From what I read in assview.c, and from all the posts, I believe what I explained in this post is correct.

Thanks.
